oidcauth/http.go

```diff
@@ -0,0 +1,98 @@
+package oidcauth
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/coreos/go-oidc"
+	"github.com/skuid/spec/middlewares"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// User is an OIDC authenticated user
+type User struct {
+	Email string `json:"email"`
+}
+
+// key is an unexported type for keys defined in this package.
+// This prevents collisions with keys defined in other packages.
+type key int
+
+// userKey is the key for user.User values in Contexts. It is
+// unexported; clients use user.NewContext and user.FromContext
+// instead of using this key directly.
+var userKey key
+
+// NewContext returns a new Context that carries value u.
+func NewContext(ctx context.Context, u *User) context.Context {
+	return context.WithValue(ctx, userKey, u)
+}
+
+// FromContext returns the User value stored in ctx, if any.
+func FromContext(ctx context.Context) (*User, bool) {
+	u, ok := ctx.Value(userKey).(*User)
+	return u, ok
+}
+
+// UserLoggingClosure adds a "user" field for an authorized user
+func UserLoggingClosure(r *http.Request) []zapcore.Field {
+	user, ok := FromContext(r.Context())
+	if !ok {
+		return []zapcore.Field{}
+	}
+	return []zapcore.Field{zap.String("user", user.Email)}
+}
+
+// OidcUserContext is a middlware for embedding a User in the request's context
+func OidcUserContext(issuerURL, clientID string) middlewares.Middleware {
+	provider, err := oidc.NewProvider(context.Background(), issuerURL)
+	if err != nil {
+		zap.L().Error("Error creating provider", zap.Error(err))
+		// painc?
```

Comment:
Leftover comment

Reply:
I'm going to leave it for now, if this is silently consuming a misconfigured server, I might want to panic here.